src/input/label.component.ts

```diff
@@ -16,77 +16,90 @@ import { TextArea } from "./text-area.directive";
+			</ng-template>
 		</div>
-		<div
-			*ngIf="!skeleton && helperText && !invalid && !warn"
-			class="bx--form__helper-text"
-			[ngClass]="{'bx--form__helper-text--disabled': disabled}">
-			<ng-container *ngIf="!isTemplate(helperText)">{{helperText}}</ng-container>
-			<ng-template *ngIf="isTemplate(helperText)" [ngTemplateOutlet]="helperText"></ng-template>
-		</div>
-		<div *ngIf="!warn && invalid" class="bx--form-requirement">
-			<ng-container *ngIf="!isTemplate(invalidText)">{{invalidText}}</ng-container>
-			<ng-template *ngIf="isTemplate(invalidText)" [ngTemplateOutlet]="invalidText"></ng-template>
-		</div>
-		<div *ngIf="!invalid && warn" class="bx--form-requirement">
-			<ng-container *ngIf="!isTemplate(warnText)">{{warnText}}</ng-container>
-			<ng-template *ngIf="isTemplate(warnText)" [ngTemplateOutlet]="warnText"></ng-template>
+		<div class="bx--text-input__field-outer-wrapper"
+			[ngClass]="{
+			'bx--text-input__field-outer-wrapper--inline': inline
+			}">
+			<div
+				[class]="wrapperClass"
+				[ngClass]="{
+					'bx--text-input__field-wrapper--warning': warn
+				}"
+				[attr.data-invalid]="(invalid ? true : null)"
+				#wrapper>
+				<svg
+					*ngIf="!warn && invalid && !hasReadonlyStyle"
+					ibmIcon="warning--filled"
+					size="16"
+					[ngClass]="{
+						'bx--text-input__invalid-icon': !textArea,
+						'bx--text-area__invalid-icon': textArea
+					}">
+				</svg>
+				<svg
+					*ngIf="!invalid && warn && !hasReadonlyStyle"
+					ibmIcon="warning--alt--filled"
+					size="16"
+					class="bx--text-input__invalid-icon bx--text-input__invalid-icon--warning">
+				</svg>
+				<svg
+					*ngIf="hasReadonlyStyle"
+					ibmIcon="edit--off"
+					size="16"
+					class="bx--text-input__readonly-icon">
+				</svg>
+				<ng-content select="input,textarea,div"></ng-content>
+			</div>
+			<ng-container *ngIf="!inline"
+				[ngTemplateOutlet]="helperTextTemplate" [ngTemplateOutletContext]="{inline: inline}">
+			</ng-container>
 		</div>
+
+		<ng-template #helperTextTemplate let-inline=inline>
+			<div *ngIf="!skeleton && helperText && ((!invalid && !warn) || inline)"
+				class="bx--form__helper-text"
+				[ngClass]="{
+					'bx--form__helper-text--disabled': disabled
+				}">
+				<ng-container *ngIf="!isTemplate(helperText)">{{helperText}}</ng-container>
+				<ng-template *ngIf="isTemplate(helperText)" [ngTemplateOutlet]="helperText"></ng-template>
+			</div>
+			<div *ngIf="!inline && !warn && invalid" class="bx--form-requirement">
+				<ng-container *ngIf="!isTemplate(invalidText)">{{invalidText}}</ng-container>
+				<ng-template *ngIf="isTemplate(invalidText)" [ngTemplateOutlet]="invalidText"></ng-template>
+			</div>
+			<div *ngIf="!inline && !invalid && warn" class="bx--form-requirement">
+				<ng-container *ngIf="!isTemplate(warnText)">{{warnText}}</ng-container>
+				<ng-template *ngIf="isTemplate(warnText)" [ngTemplateOutlet]="warnText"></ng-template>
+			</div>
+		<ng-template>
```

Comment:
I've created another branch to test this out and put it in a PR (in my fork) for viewing the diffs: https://github.com/anemonetea/carbon-components-angular/pull/1/files

(From another comment) I agree that the outer wrapper shouldn't have been getting applied to in the `textArea` variant, so my branch fixes that issue.

Also agree that the shared `ibm-label` wrapper is making this code messy and it's not the best for maintenance...

> We want to make this change in a way where we don't require existing CCA users to update their styles or make any changes. Changing order of label, adding wrapper classes will cause a breaking change for users that are making targeted style changes to label or input for example.

My concern with this approach is that we're further deviating from how React uses these CSS classes by not using `bx--text-input__field-outer-wrapper` for regular (not inline) instances of the `textInput` variant. So, if they change how the CSS works, our Angular stuff might break...

> we don't require existing CCA users to update their styles or make any changes.

At least in my experience, my product team follows the wisdom of avoiding modifications of internal Carbon classes just in case Carbon makes changes. So, it's Carbon's prerogative to make changes to the internal working of Carbon components. They're also scared of updates because they might break stuff... so, every update needs to trigger regression testing. At the end of the day, it sounds like more of a matter of setting expectations about what kind of versioning change (fix, minor, major) can be expected to make changes to the internal DOM / CSS structure of components, and which ones can't -- not sure about how you guys handle this.